### ISSUE TITLE
macOS: clients can only connect 98 times per boot

### ISSUE THREAD

**Issue by p--b** [jackaudio/jack2#784]
## Describe the bug

On macOS
If a client with a given name connects to a server with a given name 98 times
Further attempts to connect a client with that name to a server with that name will fail until the user's session is restarted (i.e. the operating system is rebooted).

## Environment

* **JACK Version:** Observed in:
`jackdmp version 1.9.18`
`jackdmp version 1.9.19`
`development` branch as of ef2af6821
I believe bug introduced in 2a4c9f74

* **Operating System:**
```
ProductName:    Mac OS X
ProductVersion: 10.15.7
BuildVersion:   19H15
```

```
ProductName:    macOS
ProductVersion: 11.4
BuildVersion:   20F71
```

* **Installation:** Built from source, also official jackd binaries 1.9.{18,19} from jackaudio.org

## Steps To Reproduce

```bash
# On macOS:
jackd -vd dummy &
for i in {1..100} ; do echo "TRIAL $i" ; jack_lsp || echo "FAILURE" ; done
```

### Expected vs. actual behavior

**Expected**: there is no limit to the number of times a client can successfully connect to jackd before the operating system must be rebooted.
**Actual**: When a client with a given name connects to a server of a given name for the 99th (or later) time, the following output is observed. Rebooting the operating system resolves the issue.

<details>
<summary>Server output</summary>

```
% jackd -vd dummy
jackdmp 1.9.20
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2021 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
JACK server starting in realtime mode with priority 10
self-connect-mode is "Don't restrict self connect requests"
Jack: JackPosixThread::StartImp : create non RT thread
Jack: JackPosixThread::ThreadHandler : start
Jack: JackDriver::Open capture_driver_name = dummy
Jack: JackDriver::Open playback_driver_name = dummy
Jack: Check protocol client = 9 server = 9
Jack: JackEngine::ClientInternalOpen: name = system
Jack: JackEngine::AllocateRefNum ref = 0
Jack: JackMachSemaphore::Allocate name = js501.system
Jack: JackEngine::NotifyAddClient: name = system
Jack: JackGraphManager::SetBufferSize size = 1024
Jack: JackConnectionManager::DirectConnect first: ref1 = 0 ref2 = 0
Jack: JackGraphManager::ConnectRefNum cur_index = 0 ref1 = 0 ref2 = 0
Jack: JackDriver::SetupDriverSync driver sem in flush mode
Jack: JackSocketServerChannel::Open
Jack: JackServerSocket::Bind : addr.sun_path /tmp/jack_default_501_0
Jack: JackSocketServerChannel::BuildPoolTable size = 1
Jack: JackEngine::Open
Jack: JackClientSocket::Connect : addr.sun_path /tmp/jack_default_501_0
Jack: JackEngine::ClientInternalOpen: name = freewheel
Jack: JackEngine::AllocateRefNum ref = 1
Jack: JackMachSemaphore::Allocate name = js501.freewheel
Jack: JackEngine::NotifyAddClient: name = freewheel
Jack: JackDriver::ClientNotify ref = 1 driver = system name = freewheel notify = 0
Jack: JackDriver::ClientNotify ref = 0 driver = freewheel name = system notify = 0
Jack: JackConnectionManager::DirectConnect first: ref1 = 1 ref2 = 1
Jack: JackGraphManager::ConnectRefNum cur_index = 0 ref1 = 1 ref2 = 1
Jack: JackDriver::SetupDriverSync driver sem in flush mode
Jack: JackAudioDriver::Attach fBufferSize = 1024 fSampleRate = 48000
Jack: JackEngine::PortRegister ref = 0 name = system:capture_1 type = 32 bit float mono audio flags = 22 buffer_size = 1024
Jack: JackGraphManager::AllocatePortAux port_index = 1 name = system:capture_1 type = 32 bit float mono audio
Jack: JackConnectionManager::AddOutputPort ref = 0 port = 1
Jack: JackEngine::ClientNotify: no callback for notification = 9
Jack: JackEngine::ClientNotify: no callback for notification = 9
Jack: JackAudioDriver::Attach fCapturePortList[i] port_index = 1
Jack: JackEngine::PortRegister ref = 0 name = system:capture_2 type = 32 bit float mono audio flags = 22 buffer_size = 1024
Jack: JackGraphManager::AllocatePortAux port_index = 2 name = system:capture_2 type = 32 bit float mono audio
Jack: JackConnectionManager::AddOutputPort ref = 0 port = 2
Jack: JackEngine::ClientNotify: no callback for notification = 9
Jack: JackEngine::ClientNotify: no callback for notification = 9
Jack: JackAudioDriver::Attach fCapturePortList[i] port_index = 2
Jack: JackEngine::PortRegister ref = 0 name = system:playback_1 type = 32 bit float mono audio flags = 21 buffer_size = 1024
Jack: JackGraphManager::AllocatePortAux port_index = 3 name = system:playback_1 type = 32 bit float mono audio
Jack: JackConnectionManager::AddInputPort ref = 0 port = 3
Jack: JackEngine::ClientNotify: no callback for notification = 9
Jack: JackEngine::ClientNotify: no callback for notification = 9
Jack: JackAudioDriver::Attach fPlaybackPortList[i] port_index = 3
Jack: JackEngine::PortRegister ref = 0 name = system:playback_2 type = 32 bit float mono audio flags = 21 buffer_size = 1024
Jack: JackGraphManager::AllocatePortAux port_index = 4 name = system:playback_2 type = 32 bit float mono audio
Jack: JackConnectionManager::AddInputPort ref = 0 port = 4
Jack: JackEngine::ClientNotify: no callback for notification = 9
Jack: JackEngine::ClientNotify: no callback for notification = 9
Jack: JackAudioDriver::Attach fPlaybackPortList[i] port_index = 4
Jack: JackServer::Start
Jack: JackThreadedDriver::Start
Jack: JackPosixThread::StartImp : create non RT thread
Jack: JackPosixThread::ThreadHandler : start
Jack: JackThreadedDriver::Init real-time
Jack: JackMachThread::AcquireSelfRealTime fPeriod = 21333 fComputation = 100 fConstraint = 21333
Jack: JackMachThread::thread_policy_set res = 0
Jack: JackPosixThread::StartImp : create non RT thread
Jack: JackPosixThread::ThreadHandler : start
Jack: JackSocketServerChannel::ClientCreate socket
Jack: JackSocketServerChannel::BuildPoolTable size = 2
Jack: JackSocketServerChannel::BuildPoolTable fSocketTable i = 1 fd = 11
Jack: JackRequest::Notification
Jack: JackDriver::ClientNotify ref = 1 driver = freewheel name = freewheel notify = 18
Jack: JackDriver::ClientNotify ref = 1 driver = freewheel name = freewheel notify = 18
Jack: JackDriver::ClientNotify ref = 1 driver = freewheel name = freewheel notify = 18
Jack: JackDriver::ClientNotify ref = 1 driver = freewheel name = freewheel notify = 18
Jack: JackEngine::ClientNotify: no callback for notification = 4
Jack: JackEngine::ClientNotify: no callback for notification = 4
Jack: JackSocketServerChannel::ClientCreate socket
Jack: JackSocketServerChannel::BuildPoolTable size = 3
Jack: JackSocketServerChannel::BuildPoolTable fSocketTable i = 1 fd = 11
Jack: JackSocketServerChannel::BuildPoolTable fSocketTable i = 2 fd = 12
Jack: JackSocketServerChannel::Execute : poll client error err = Undefined error: 0
Jack: JackSocketServerChannel::ClientKill ref = -1 fd = 12
Jack: Client was not opened : probably correspond to server_check
Jack: JackClientSocket::Close
Jack: JackSocketServerChannel::BuildPoolTable size = 2
Jack: JackSocketServerChannel::BuildPoolTable fSocketTable i = 1 fd = 11
Jack: JackSocketServerChannel::ClientCreate socket
Jack: JackSocketServerChannel::BuildPoolTable size = 3
Jack: JackSocketServerChannel::BuildPoolTable fSocketTable i = 1 fd = 11
Jack: JackSocketServerChannel::BuildPoolTable fSocketTable i = 2 fd = 12
Jack: JackRequest::ClientCheck
Jack: Check protocol client = 9 server = 9
Jack: JackRequest::ClientOpen
Jack: JackEngine::ClientExternalOpen: uuid = 3, name = lsp
Jack: JackEngine::AllocateRefNum ref = 2
Jack: JackMachSemaphore::Allocate name = js501.lsp
Cannot allocate synchro
Jack: JackMachSemaphore::Destroy name = js501.lsp
Jack: JackExternalClient::Close
Jack: JackSocketNotifyChannel::Close
Jack: JackClientSocket::Close
Cannot create new client
Jack: JackSocketServerChannel::Execute : poll client error err = Socket is not connected
Jack: JackSocketServerChannel::ClientKill ref = -1 fd = 12
Jack: Client was not opened : probably correspond to server_check
Jack: JackClientSocket::Close
Jack: JackSocketServerChannel::BuildPoolTable size = 2
Jack: JackSocketServerChannel::BuildPoolTable fSocketTable i = 1 fd = 11
```

</details>

<details>
<summary>Client output</summary>

```
% jack_lsp 
Cannot open lsp client
JackShmReadWritePtr1::~JackShmReadWritePtr1 - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Error: cannot connect to JACK, jack_client_open() failed, status = 0x21
```

</details>

## Analysis
I believe this happens because `JackMachSemaphore` communicates the mach port for the semaphore to the client using a dynamically generated system-global resource, a named service created via `bootstrap_register`. This is done by `JackMachSemaphore::recursiveBootstrapRegister(...)`.

The `recursive` is needed because the bootstrap server will refuse to allow us to re-register a previously-used service name (and also will not allow us to re-use a previously-created service either, I believe as our bootstrap port is not privileged and for some reason launchd thinks the dead service ports are still active?).

As such, in the event of most kinds of failure, we simply try again, appending an incrementing integer to the end of the service name. There does not seem to be any means available to us that I can find of causing the bootstrap server to forget about a previously-registered service, meaning that once registered, these service registrations persist until the user's session is restarted (i.e. reboot). We can verify this by running `jackd`, executing `jack_lsp` a few times, and inspecting the list of registered services:

<details>
<summary>launchctl print output showing registered service ports persisting</summary>

```
% launchctl print user/`id -u`
<snip>
        unmanaged processes = {
                com.apple.xpc.launchd.unmanaged.jackd.496 = {
                        active count = 13
                        dynamic endpoints = {
                                "js501.lsp-10" = {
                                        port = 0x7582b
                                        active = 1
                                        managed = 0
                                        reset = 0
                                        hide = 0
                                }
                                "js501.lsp-9" = {
                                        port = 0x7a01b
                                        active = 1
                                        managed = 0
                                        reset = 0
                                        hide = 0
                                }
                                "js501.lsp-8" = {
                                        port = 0x7c037
                                        active = 1
                                        managed = 0
                                        reset = 0
                                        hide = 0
                                }
                                "js501.lsp-7" = {
                                        port = 0x79f1f
                                        active = 1
                                        managed = 0
                                        reset = 0
                                        hide = 0
                                }
                                "js501.lsp-6" = {
                                        port = 0x7f85f
                                        active = 1
                                        managed = 0
                                        reset = 0
                                        hide = 0
                                }
                                "js501.lsp-5" = {
                                        port = 0x7a3fb
                                        active = 1
                                        managed = 0
                                        reset = 0
                                        hide = 0
                                }
                                "js501.lsp-4" = {
                                        port = 0x7a23b
                                        active = 1
                                        managed = 0
                                        reset = 0
                                        hide = 0
                                }
                                "js501.lsp-3" = {
                                        port = 0x801bb
                                        active = 1
                                        managed = 0
                                        reset = 0
                                        hide = 0
                                }
                                "js501.lsp-2" = {
                                        port = 0x79b17
                                        active = 1
                                        managed = 0
                                        reset = 0
                                        hide = 0
                                }
                                "js501.lsp" = {
                                        port = 0x79c1b
                                        active = 1
                                        managed = 0
                                        reset = 0
                                        hide = 0
                                }
                                "js501.freewheel" = {
                                        port = 0x75907
                                        active = 1
                                        managed = 0
                                        reset = 0
                                        hide = 0
                                }
                                "js501.system" = {
                                        port = 0x159f
                                        active = 1
                                        managed = 0
                                        reset = 0
                                        hide = 0
                                }
                        }
<snip>
                }
        }
<snip>
```

</details>

Unfortunately, `recursiveBootstrapRegister(...)` has a [recursion limit](https://github.com/jackaudio/jack2/blob/852ba2b8ebc944865e75d8eb5d4550d408772f3b/macosx/JackMachSemaphore.mm#L114) which fails silently if it is hit. Investigation reveals that this is the cause of this failure; each time the client is run, another service port is created, which persist until there are 98, after which `recursiveBootstrapRegister` declines to create any more and semaphore creation fails.

## Proposed Fix
Disclaimer: I am no mach/darwin wizard! However, investigating `bootstrap_register(...)` in [bootstrap.h](https://opensource.apple.com/source/launchd/launchd-328/launchd/src/bootstrap.h.auto.html), the documentation suggests that this approach of registering a service port via the bootstrap server to establish IPC is not recommended (and is actually now deprecated). Our use-case seems to match exactly the described case (3):

```
 * 3) If the code was registering a dynamically generated string and passing
 *    the string to other applications, please rewrite the code to send a Mach
 *    send-right directly.
```

As we are already creating a named shared memory segment to communicate the service name to the client, we could instead simply use the same shared memory segment to pass the send right for the semaphore directly instead, obviating any need to register services with the bootstrap server at all.

I have implemented this change and it seems to resolve the issue; I will create a PR shortly.

**Comment by falkTX** [comment]
changes/patches welcome. I got this to work besides with the limitations you mentioned. Tried some cleaning up of the bootstrap stuff but couldnt figure out how to do that (maybe not possible?)

please note that this semaphore setup needs to work across architectures. this is what makes it so tricky.
the semaphores (on shared memory) need to be usable from both x64 and arm64 modes.

**Comment by HaHeho** [comment]
I believe this is was was also reported in https://github.com/jackaudio/jack2/issues/658 (and e.g. https://github.com/spatialaudio/jackclient-python/issues/98).

But thanks a lot for your investigation and the fix (hopefully). <3

**Comment by falkTX** [comment]
nah, this one is very much macOS specific. there is a hardcoded 99 limit for the clients, as to not let things get out of hand (too many global port registrations that cannot be cleaned up)

**Comment by p--b** [comment]
NB if anyone looks at this in future -the proposed fix I discussed above (#785) didn't quite work, so we did #788 instead.
